### PR TITLE
Migrate tests to pytest

### PR DESCRIPTION
**Pull request by emilazy**
Hi there. The nose Python test framework has been deprecated and unmaintained for over a decade, and doesn’t work out of the box on Python 3.12. Over in NixOS we’ve been working to remove it and migrating packages as we go. I’ve done an automatic migration to the modern pytest framework here with the `nose2pytest` tool, and cleaned up the remaining references manually.

Unfortunately and ironically, I can’t adequately test this PR as the `mhcflurry` package is broken in NixOS for unrelated TensorFlow reasons right now, but I hope what I have here can be useful and if you run into any issues testing it I’m happy to help fix them :) Thanks for your work on such an incredible project!

### CODE REVIEW

**Comment by emilazy** [comment]
Apologies; as expected my untested changes are broken. If you’re interested in the migration I can try to fix it up, even though I don’t have an easy way to test it end‐to‐end right now.

File: test/test_class1_affinity_predictor.py
Comment (ndalchau) [inline test/test_class1_affinity_predictor.py:16]: The CI is failing because you've removed `assert_raises` here, yet the function is still being called in this test file.

**Comment by ndalchau** [comment]
I think this is a nice addition, as it will help maintain this version of MHCflurry into the future. But it's up to @timodonnell 
 to approve :)

It should be an easy fix to get the CI running (see my comment above). What is the challenge behind having an end-to-end  test for this repo? I didn't have a problem running this locally.

**Comment by emilazy** [comment]
Apologies for the untested PR :) `nose2pytest` is somewhat unsophisticated so it missed a few things and mangled some syntax. I’ve pushed fixes for the ones reported by CI; hopefully it will be happy now.

The reason I sent this without testing is that the MHCflurry package in Nixpkgs is currently broken due to our out‐of‐date TensorFlow, and we don’t currently support TensorFlow on macOS, which meant I couldn’t get a proper build environment going to test this. I use Nixpkgs for all my package management, so it’d be somewhat awkward to install a bunch of stuff outside of that for a codebase I was doing drive‐by fixes for. I realize that playing CI whack‐a‐mole when I can’t easily directly test my changes isn’t great, though… just didn’t want to leave my branch rotting on my disk in case it could be helpful to you.

**Comment by emilazy** [comment]
I’m not sure what’s up with the latest test failures, but from what I can see they don’t relate to the nose → pytest conversion. Maybe a NumPy version compatibility issue or something?

**Comment by ndalchau** [comment]
I just took a look on my system. You're quite right that there is a problem with numpy. The last successful run of the CI brought down numpy version 1.26.4: https://github.com/openvax/mhcflurry/actions/runs/11137285619.
In this PR, the CI brought down version 2.0.2. I'm pretty sure this is the problem. We need to raise a new issue for this.

**Comment by emilazy** [comment]
Ah, yes – a lot of packages aren’t ready for NumPy 2 yet, so you’re in good company :) There’s a [migration guide](https://numpy.org/doc/stable/numpy_2_0_migration_guide.html) you might find helpful, but the simplest solution for now is probably to add an upper version bound to whichever of your dependencies is pulling NumPy in so that CI works again. Happy to rebase this PR if that’s done so that the changes to the test suite can be verified.

**Comment by ndalchau** [comment]
Can you try rebasing please?

**Comment by emilazy** [comment]
Done :) Thanks for your patience with this!

**Comment by ndalchau** [comment]
@timodonnell , the CI is awaiting approval

**Comment by emilazy** [comment]
Looks like the conversion tool couldn’t figure out when it was `numpy.testing.assert_equal` instead of nose’s. Hopefully this should fix the last few failing tests!

**Review by ndalchau — APPROVED** [review]

**Comment by ndalchau** [comment]
Looks good to me. Over to the maintainers now.

**Comment by timodonnell** [comment]
Thanks @emilazy and @ndalchau for all of your work on this! Looks great to me. Very much appreciated.

**Comment by GaetanLepage** [comment]
Indeed, thanks all for this great patch that should help us fixing `mhcflurry` in nixpkgs.
@timodonnell do you plan to have a release (that would contain this patch) any time soon ?
If not we will vendor the patch ourselves in the meantime.

**Pull request merged**
